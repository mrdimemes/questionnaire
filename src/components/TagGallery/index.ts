export { default as TagGallery } from "./TagGallery";
