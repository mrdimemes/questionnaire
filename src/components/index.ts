export { App } from "./App";

export { Header } from "./header";

export { PaginationBar } from "./pagination";

export { TagComponent } from "./tags";
export { TagGallery } from "./tags";
export { TagBar } from "./tags";
export { AddTagWidget } from "./tags";

export { QuestionnaireGallery } from "./questionnaires/gallery";
export { QuestionnaireComponent } from "./questionnaires/questionnaire";
export { QuestionnaireEditor } from "./questionnaires/edit";

export { Button } from "./UI/Button";
export { Input } from "./UI/inputs";
export { CheckboxInput } from "./UI/inputs";
export { RadioInput } from "./UI/inputs";
export { TextInput } from "./UI/inputs";
export { EmailInput } from "./UI/inputs";
export { PasswordInput } from "./UI/inputs";
export { NameInput } from "./UI/inputs";
export { TextArea } from "./UI/inputs";
export { Select } from "./UI/inputs";

export { AuthForm } from "./AuthForm";

export { LoadingSpinner } from "./LoadingSpinner";