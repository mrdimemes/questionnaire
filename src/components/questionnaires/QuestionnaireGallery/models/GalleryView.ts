export enum GalleryView {
  Rows = "ROWS",
  Plates = "PLATES",
};