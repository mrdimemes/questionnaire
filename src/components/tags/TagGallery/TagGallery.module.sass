.body
  max-height: 80vh
  display: flex
  flex-direction: column
  flex-wrap: wrap
  gap: 15px
  justify-content: center
  position: relative
  padding-bottom: 100px

.pagination
  position: absolute
  bottom: 0
  left: 0
  right: 0