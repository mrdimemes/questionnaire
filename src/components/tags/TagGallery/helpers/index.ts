export { default as getCompareFunction } from "./getCompareFunction";
export { default as searchFilter } from "./searchFilter";