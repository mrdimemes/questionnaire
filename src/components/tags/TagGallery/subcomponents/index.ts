export { TagGalleryItem } from "./TagGalleryItem";
export { AddTagWidget } from "./AddTagWidget";
export { AddTagForm } from "./AddTagForm";