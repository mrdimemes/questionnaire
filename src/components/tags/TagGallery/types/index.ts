export type { TagGalleryProps } from "./TagGalleryProps";