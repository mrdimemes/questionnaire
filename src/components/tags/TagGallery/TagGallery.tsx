import { useState, useEffect } from "react";
import { useTagsDataSelector, useAppDispatch } from "src/redux/hooks";
import { removeTag as removeTagAction } from "src/redux/slices/tagsSlice";
import { LoadingSpinner, PaginationBar, Button } from "src/components";
import { TagComponent, AddTagWidget } from "../";
import { FetchStatus, Tag } from "src/models";
import styles from "./TagGallery.module.sass";
import { QuestionnaireService } from "src/services";
import { FetchError } from "src/api/errors";


type TagGalleryProps = {
  resetScroll?: () => void;
}

const TagGallery = ({ resetScroll }: TagGalleryProps) => {
  const [tags, status] = useTagsDataSelector();
  const tagsPerPage = 20;
  const totalPages = Math.ceil(tags.length / tagsPerPage);
  const [activePage, setActivePage] = useState(1);
  const startIndex = (activePage - 1) * tagsPerPage;
  const dispatch = useAppDispatch();

  const removeTag = (tag: Tag) => {
    const alarmText = `Вы дейсвительно хотите удалить тэг "${tag.label}" ` +
      `(опросов: ${tag.freq})?`;
    const confirmation = window.confirm(alarmText);
    if (confirmation) {
      QuestionnaireService.removeTag(tag.id).then(response => {
        if (response instanceof FetchError) {
          console.log(response)
        } else {
          dispatch(removeTagAction(tag.id));
        }
      });
    }
  };

  useEffect(() => {
    if (resetScroll) resetScroll();
  }, [activePage, resetScroll]);

  return (
    <div className={styles.gallery}>
      {status === FetchStatus.Loading && <LoadingSpinner />}

      {status === FetchStatus.Complete && <>
        <AddTagWidget className={styles.addTagWidget} />
        {tags.slice(startIndex, startIndex + tagsPerPage).map((tag) => {
          return (
            <div
              className={styles.tagContainer}
              key={`tag-container-${tag.id}`}
            >
              <TagComponent
                key={`tag-${tag.id}`}
                label={tag.label}
                freq={tag.freq}
              />
              <Button
                className={styles.removeTagButton}
                key={`tag-rb-${tag.id}`}
                onClick={() => removeTag(tag)}
                children="x"
              />
            </div>
          )

        })}
      </>}

      <PaginationBar
        className={styles.pagination}
        activePage={activePage}
        totalPages={totalPages}
        callback={setActivePage}
      />

    </div>
  );
}

export default TagGallery