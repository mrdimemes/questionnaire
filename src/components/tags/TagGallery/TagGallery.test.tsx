import { screen } from "@testing-library/react";
import { renderWithProviders } from "src/utils/test-utils";
import { setupStore } from "src/redux";

import { FetchStatus } from "src/models";

import TagGallery from "./TagGallery";

describe("TagGallery component", () => {
  test("TagGallery renders", () => {
    const store = setupStore({
      tags: {
        tags: [
          { id: 0, label: "tag0", frequency: 99 },
          { id: 1, label: "tag1", frequency: 1 },
          { id: 2, label: "tag2", frequency: 205 }],
        tagsLoadingStatus: FetchStatus.Complete,
      },
    });
    renderWithProviders(<TagGallery />, { store });
    expect(screen.getByText("tag0")).toBeInTheDocument();
    expect(screen.getAllByText(/tag/i)).toHaveLength(3);
    expect(screen.getByText("99")).toBeInTheDocument();
    expect(screen.getByText("1")).toBeInTheDocument();
    expect(screen.getByText("205")).toBeInTheDocument();
  });

  test("TagGallery empty snapshot", () => {
    const store = setupStore({
      tags:
        { tags: [], tagsLoadingStatus: FetchStatus.Complete },
    });
    const { container } = renderWithProviders(<TagGallery />, { store });
    expect(container).toMatchSnapshot();
  });

  test("TagGallery fillfull snapshot", () => {
    const store = setupStore({
      tags: {
        tags: [
          { id: 0, label: "tag0", frequency: 99 },
          { id: 1, label: "tag1", frequency: 1 },
          { id: 2, label: "tag2", frequency: 205 }],
        tagsLoadingStatus: FetchStatus.Complete,
      },
    });
    const { container } = renderWithProviders(<TagGallery />, { store });
    expect(container).toMatchSnapshot();
  });
});