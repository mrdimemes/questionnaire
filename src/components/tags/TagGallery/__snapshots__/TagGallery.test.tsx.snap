// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagGallery component TagGallery empty snapshot 1`] = `
<div>
  <div
    class="gallery"
  >
    <div
      class="widget addTagWidget"
    >
      <button
        class="body theme_light focusButton"
      >
        +
      </button>
    </div>
    <div
      class="Bar pagination"
    >
       
       
       
      <button
        class="body theme_light"
      >
        Вперёд
      </button>
       
    </div>
  </div>
</div>
`;

exports[`TagGallery component TagGallery fillfull snapshot 1`] = `
<div>
  <div
    class="gallery"
  >
    <div
      class="widget addTagWidget"
    >
      <button
        class="body theme_light focusButton"
      >
        +
      </button>
    </div>
    <div
      class="tagContainer"
    >
      <div
        class="body theme_light"
      >
        <div>
          tag0
        </div>
        <div
          class="freq freq_theme_light"
        >
          99
        </div>
      </div>
      <button
        class="body theme_light removeTagButton"
      >
        x
      </button>
    </div>
    <div
      class="tagContainer"
    >
      <div
        class="body theme_light"
      >
        <div>
          tag1
        </div>
        <div
          class="freq freq_theme_light"
        >
          1
        </div>
      </div>
      <button
        class="body theme_light removeTagButton"
      >
        x
      </button>
    </div>
    <div
      class="tagContainer"
    >
      <div
        class="body theme_light"
      >
        <div>
          tag2
        </div>
        <div
          class="freq freq_theme_light"
        >
          205
        </div>
      </div>
      <button
        class="body theme_light removeTagButton"
      >
        x
      </button>
    </div>
  </div>
</div>
`;
